Store digests of config files in static Pods annotations

gdemonet opened the pull request:
**Component**: salt, kubernetes
**Type**: bugfix

**Context**:

- In order to have `kubelet` restart static Pod containers on configuration changes, we need to actually change the static Pod manifest

**Summary**:

- Introduce a basic `metalk8s.static_pod_managed` helper state to compute config files digests at runtime (so we can still rely on Salt `require`/`watch` ordering to lazily evaluate the hashes)
    - Hashing algorithm is basically: `md5 ( '-'.join( sha256(file for file in config_files) )`

- Use it for Salt master
- Use it for K8s control plane components (`k-a`, `k-s`, `k-c-m`) and `etcd`
- Use it for the package repositories and OCI registry

- Each static pod includes **every** single file that it wants to read (certificates, private keys, kubeconfig, etc.)

**Acceptance criteria**:

- On changes to the Salt master configuration files (**made using Salt**), `salt-master` Pod should be "restarted" by `kubelet` (actually replaced)
- On changes to other config files (e.g. certificates, kubeconfigs), related Pods should be "restarted"

**Notes**:

When trying to make this PR get a green build, I ran into many CI issues linked to "not waiting long enough" in tests. In the process of fixing this, a few modifications around tests were made:

- Pytest markers are now documented and used by default: see a4f1f09ae0ff04e78c6c59fe0becfe23f4ddf679
- The "bootstrap idempotent" test only runs the bootstrap once, since it's a `@post` test, and is now marked as `@slow`, which runs after the other tests in the CI (and can be filtered out easily with `tox -e tests -- -m "not slow"`)
- A number of tests now wait (through multiple retries) much longer than they used to before attempting operations on `Running` Pods

---

Fixes: GH-937
Fixes: GH-881

gdemonet committed:
Define basic 'file.managed' wrapper state

We want to compute config file digests at runtime, so that the Salt
require/watch ordering rules can kick in first.
A quick solution was to wrap 'file.managed' inside a custom
'metalk8s.static_pod_managed' state, which computes a list of SHA256
digests from a list of file paths, and forwards it as a string to the
Pod manifest Jinja template.

Issue: GH-937

gdemonet committed:
Use 'static_pod_managed' for Salt master

Issue: GH-937

gdemonet committed:
Watch Kubernetes control-plane config changes

Issue: GH-937

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/937-restart-static-pods` into
`development/2.0`
* [w/2.1/bugfix/937-restart-static-pods](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/bugfix/937-restart-static-pods) will be merged into `development/2.1` 
* [w/2.2/bugfix/937-restart-static-pods](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/bugfix/937-restart-static-pods) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

NicolasT commented on salt/_states/metalk8s.py:
```diff
@@ -0,0 +1,32 @@
+        template="jinja",
+        user="root",
+        group="root",
+        mode="0644",
```
Maybe this should be `0600`? No need to be world-readable.

gdemonet replied:
Ok will do

NicolasT commented on salt/_states/metalk8s.py:
```diff
@@ -0,0 +1,32 @@
+def static_pod_managed(name, source, config_files, context):
+    """Simple helper to edit a static Pod manifest if configuration changes.
+
+    Expects the template to use the `config_digests` variable and store it in
```
Why plural?

NicolasT commented on salt/_states/metalk8s.py:
```diff
@@ -0,0 +1,32 @@
+        mode="0644",
+        makedirs=False,
+        backup=False,
+        context=dict(context, config_digests=",".join(config_digests)),
```
Uh, nah. That'd mean we can get some very long but otherwise useless annotation on the Pod. If indeed we want to use `sha256` (which isn't even required here), then the result should be, I don't know, a checksum of the concatenation of all sorted checksums or something? Basically, one single value, ideally short, which is (quasi-)guaranteed to change when the input(s) change.

gdemonet replied:
Would something along the lines of `md5( '-'.join( sha256(file) for file in config_files )` suit our needs ?

gdemonet replied:
Done

NicolasT commented on salt/metalk8s/kubernetes/apiserver/installed.sls:
```diff
@@ -80,15 +80,13 @@ Set up default basic auth htpasswd:
+    - config_files:
+        - {{ htpasswd_path }}
+        - /etc/keepalived/check-apiserver.sh
+        - /etc/keepalived/keepalived.conf.sh
```
Also API-server certificates (CA, server cert, `kubelet` client cert,...)

gdemonet replied:
OK, I was wondering about certs

gdemonet replied:
What about public/private keys? I added `sa.pub` for `k-a` but was not sure whether to store the private key digest (security?)...

NicolasT replied:
Let's assume the digest function is strong enough not to be reversible... It's the MD5 of a concatenation of SHA256s of the inputs, should be fine ;)

NicolasT commented on salt/metalk8s/kubernetes/apiserver/installed.sls:
```diff
@@ -80,15 +80,13 @@ Set up default basic auth htpasswd:
-    - dir_mode: 750
+    - config_files:
+        - {{ htpasswd_path }}
+        - /etc/keepalived/check-apiserver.sh
```
These only when `keepalived` is enabled, right?

gdemonet replied:
Done

NicolasT commented on salt/metalk8s/kubernetes/controller-manager/installed.sls:
```diff
@@ -45,5 +41,5 @@ Create kube-controller-manager Pod manifest:
             name: kubeconfig
             type: File
-    - require:
+    - onchanges:
```
Also a requisite, no?

gdemonet replied:
So, from what I understand, `onchanges` behaves exactly like a `require` if the referenced state does not return information about changes (so only executes if the requisite returned `True`), otherwise looks at the changes to see if needs to execute. Ordering should be taken care of nicely, if that was your concern.

NicolasT commented on salt/metalk8s/kubernetes/files/control-plane-manifest.yaml:
```diff
@@ -3,6 +3,7 @@ kind: Pod
 metadata:
   annotations:
     scheduler.alpha.kubernetes.io/critical-pod: ""
+    metalk8s.scality.com/config-digests: "{{ config_digests }}"
```
Could be singular, if aforementioned changes are made.

NicolasT commented on salt/metalk8s/kubernetes/scheduler/installed.sls:
```diff
@@ -29,5 +25,5 @@ Create kube-scheduler Pod manifest:
             name: kubeconfig
             type: File
-    - require:
+    - onchanges:
```
And require?

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

bert-e commented:
# History mismatch



Merge commit #ac15704acb9919312c56315b82a359db903478aa on the integration branch
`w/2.1/bugfix/937-restart-static-pods` is merging a branch which is neither the current
branch `bugfix/937-restart-static-pods` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `bugfix/937-restart-static-pods` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
@NicolasT I changed the digest computation (only store in the annotation the MD5 sum of the concatenation of SHA256 digests of listed `config_files`), and added certificates where relevant.

Now, I'm wondering if this would warrant some automatic tests, or, since it's mostly about upgrade/development scenarii, if we can simply check it manually. If the former, how would we do such tests? I'm thinking about simply testing the `metalk8s.static_pod_managed` state with some arbitrary static Pod (not a useful one, just an artificial one for the test) to restart based on some arbitrary file... Would that work? Or do we want to check that, for each service, configuration changes effectively get picked up by said service?

NicolasT commented on salt/metalk8s/kubernetes/scheduler/installed.sls:
```diff
@@ -4,15 +4,11 @@ include:
-    - makedirs: True
-    - dir_mode: 750
+    - config_files:
+      - /etc/kubernetes/scheduler.conf
```
Shouldn't this one have `/etc/kubernetes/pki/ca.crt` just like `k-c-m`?

gdemonet replied:
It does not depend on it, neither does it mount the `/etc/kubernetes/pki` directory, so I guess not? In any case, the certificate data is embedded in the generated kubeconfig, and the `metalk8s_kubeconfig: Create kubeconfig file for scheduler` state retrieves the "latest" certificate data from the mine.

NicolasT replied:
Hmh, then why would `k-c-m` require `ca.crt`?

gdemonet replied:
I put it there because it mounts `/etc/kubernetes/pki` and has a flag for this cert `--cluster-signing-cert-file=/etc/kubernetes/pki/ca.crt`, but maybe it's not required indeed.

charlesprost commented:
> Now, I'm wondering if this would warrant some automatic tests, or, since it's mostly about upgrade/development scenarii, if we can simply check it manually. If the former, how would we do such tests? I'm thinking about simply testing the `metalk8s.static_pod_managed` state with some arbitrary static Pod (not a useful one, just an artificial one for the test) to restart based on some arbitrary file... Would that work? Or do we want to check that, for each service, configuration changes effectively get picked up by said service?

Yeah I'd say it would be good to test this feature with a dummy pod. Let's have a pod doing nothing, based on a configuration file with some random string, let's change this string, and see if the pod restarts as expected.

Regarding the image, we could reuse a pod such as nginx without mounting anything.

gdemonet committed:
Watch etcd config changes

Issue: GH-937

gdemonet committed:
Test the static_pod_managed custom state

Issue: GH-937

gdemonet committed:
Increase waiting time for running Pods in tests

We had some flaky situations where waiting 15s would not be enough for
having a running Pod after its creation in K8s API.

gdemonet committed:
Improve "bootstrap idempotent" test

Since this test is run post-install, running it twice is unnecessary.

Issue: GH-881

gdemonet committed:
Document and use pytest markers

We document the existing @ci, @local, and @post markers for feature
files. One can run "pytest --markers" to see a list of registered
markers.
We use the "-m local" filter by default in tox environments, and switch
to "-m ci" when running tests in the CI.
We also introduce a @slow marker, and apply it to the "Bootstrap"
feature.
We then split CI test execution in two, running with "-m 'not slow'"
first, then "-m slow". It should help providing feedback faster to
developers.

NicolasT commented on tests/conftest.py:
```diff
@@ -10,3 +14,43 @@ def pytest_addoption(parser):
+
+
+@pytest.fixture(scope="module")
+def version(request, host):
```
`short_version`?

gdemonet replied:
I just moved it, but yes let's rename it :)

gdemonet committed:
Add "wait_for_pod" helper

This can be used in both "static Pod" and "DNS" tests. We also make sure
to name retried operations for easier debugging.

Issue: GH-937

NicolasT commented on tests/conftest.py:
```diff
@@ -10,3 +14,43 @@ def pytest_addoption(parser):
+
+@pytest.fixture
+def k8s_client(kubeconfig):
+    kubernetes.config.load_kube_config(config_file=kubeconfig)
```
This relies on the 'global' config object in the K8s Python API, which can cause difficult-to-debug issues if we ever want to test RBAC authz and such.

Consider using `new_client_from_config`, then construct an API access object passing in the returned client explicitly.

Also, this actually returns a corev1 API client, not an `ApiClient`, which I'd expect given the name of the procedure.

gdemonet replied:
Sure, will do!

gdemonet replied:
So I still have a `k8s_client` for clients exposing a single API, and `k8s_apiclient` that returns an `ApiClient` to use in `k8s_client`. By default, we use `CoreV1Api`, but it can be parametrized with `pytest.mark.parametrize`.

gdemonet committed:
Rename "version" fixture to "short_version"

gdemonet committed:
Split fixture for generating a K8s API client

We also change the approach for reading configuration, from a global
attribute in "kubernetes.config", to a local "Configuration" object.

Issue: GH-937

NicolasT reviewed (dismissed):
:ship:

gdemonet commented:
> :ship:

Sadly, I'm having weird results in CI tests execution, so your approval was dismissed @NicolasT - will ping once it's reliably fixed.

gdemonet committed:
Use "static_pod_managed" for nginx repository

Issue: GH-937

gdemonet committed:
Use "static_pod_managed" for Docker registry

Issue: GH-937

gdemonet commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/937-restart-static-pods` into
`development/2.0`
* [w/2.1/bugfix/937-restart-static-pods](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/bugfix/937-restart-static-pods) will be merged into `development/2.1` 
* [w/2.2/bugfix/937-restart-static-pods](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/bugfix/937-restart-static-pods) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @jbertran

* @monpote

* @alxf

* @gdemonet

* @lucieleonard

* @zempashi

* @carlito-scality

* @alexandre-allard-scality

* @nootal

* @nicolast

* @thomasdanan

* @tcarmet

* @slaperche-scality

* @teddyandrieux

* @sayf-eddine-scality

* @ramanduh

gdemonet commented:
Finally, three green builds! :shipit: :)

Reviewers should note the PR description changed, and the changeset now includes improvements to the test suite (especially #881, but also around `pytest` markers, that one can discover using `tox -e tests -- --markers` for instance).

gdemonet commented:
@bert-e approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @jbertran

* @monpote

* @alxf

* @gdemonet

* @lucieleonard

* @zempashi

* @carlito-scality

* @alexandre-allard-scality

* @nootal

* @nicolast

* @thomasdanan

* @tcarmet

* @slaperche-scality

* @teddyandrieux

* @sayf-eddine-scality

* @ramanduh










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye gdemonet.
